src/state_machine.zig

```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
     fn create_transfer(self: *StateMachine, t: Transfer) CreateTransferResult {
         assert(t.timestamp > self.commit_timestamp);
 
-        if (t.flags.padding != 0) return .reserved_flag_padding;
-        if (t.flags.two_phase_commit) {
-            // Otherwise reserved amounts may never be released:
-            if (t.timeout == 0) return .two_phase_commit_must_timeout;
-        } else if (t.timeout != 0) {
-            return .timeout_reserved_for_two_phase_commit;
-        }
-        if (!t.flags.condition and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+        // Either a 2-phase transfer post/void
+        if (t.flags.post_pending_transfer and t.flags.void_pending_transfer) {
+            return .cannot_post_and_void_pending_transfer;
+        } else if (t.flags.post_pending_transfer or t.flags.void_pending_transfer) {
+            if (!t.flags.hashlock and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+            if (t.flags.padding != 0) return .reserved_flag_padding;
 
-        if (t.amount == 0) return .amount_is_zero;
+            const lookup = self.get_transfer(t.id) orelse return .transfer_not_found;
+            assert(t.timestamp > lookup.timestamp);
 
-        if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+            if (!lookup.flags.pending) return .transfer_not_pending;
 
-        // The etymology of the DR and CR abbreviations for debit and credit is interesting, either:
-        // 1. derived from the Latin past participles of debitum and creditum, debere and credere,
-        // 2. standing for debit record and credit record, or
-        // 3. relating to debtor and creditor.
-        // We use them to distinguish between `cr` (credit account), and `c` (commit).
-        var dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
-        var cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
-        assert(t.timestamp > dr.timestamp);
-        assert(t.timestamp > cr.timestamp);
+            if (self.get_commit(t.id)) |exists| {
+                if (!exists.flags.void_pending_transfer and t.flags.void_pending_transfer) return .transfer_already_posted;
+                if (exists.flags.void_pending_transfer and !t.flags.void_pending_transfer) return .transfer_already_voided;
+                return .transfer_already_posted;
+            }
 
-        if (dr.unit != cr.unit) return .accounts_have_different_units;
```

Comment:
We should still rename `unit` to `ledger` and add `ledger` to Transfer as part of this PR.

Then here, if the `ledger` value is non-zero, we must check that it is the same as that of the pending transfer. Same for `code`.

Reply:
Resolved.